src/main.rs

```diff
@@ -0,0 +1,59 @@
+use crate::parser::parse_html_to_lei;
+use std::collections::HashMap;
+use std::env;
+use std::fs::File;
+use std::io::Write;
+use walkdir::{DirEntry, WalkDir};
+
+mod parser;
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let folder_path = &args[1]; // TODO: error handler
+
+    let walker = WalkDir::new(folder_path).into_iter();
+
+    let mut directories = HashMap::new();
+    let mut current_directory = String::new();
+    let mut leis = Vec::new();
+    for entry in walker.filter_entry(|entry| is_not_hidden(entry)).skip(1) {
+        let entry = entry.unwrap();
+
+        if entry.file_type().is_dir() {
+            current_directory = entry.file_name().to_os_string().into_string().unwrap();
+
+            directories.insert(current_directory.clone(), 0);
+        } else if entry.file_type().is_file()
+            && entry.file_name().to_string_lossy().ends_with(".html")
+        {
+            let lei = parse_html_to_lei(
+                entry.path().to_str().unwrap(), // TODO: handle error
+                current_directory.to_string(),
+            );
+            let lei_to_serialize = serde_json::to_string_pretty(&lei).unwrap();
+            // TODO: não usar vector para armazenar as leis
+            leis.push(lei_to_serialize);
+
+            *directories.get_mut(&current_directory).unwrap() += 1;
+        }
+    }
+
+    // TODO: escrever em formato de tabela igual no futiba
+    for (directory, files_number) in &directories {
+        println!("diretório {}: {} arquivos lidos", directory, files_number);
+    }
+
+    let data = leis.join(",");
+    let mut leis_file = File::create("leis.json").expect("Unable to create file");
+    leis_file
+        .write_all(format!("[{}]", data).as_bytes())
```

Comment:
Pq não usar o `serde` pra serializar o array inteiro? Vi que os items são derive serialize já.

Algo do tipo:

```suggestion
        .write_all(&leis.as_str())
```

Assim você aproveita que `&str` vira `&[u8]` implicitamente, e não precisa do format.

Reply:
Melhor ainda, usando o `writer` trait pra um arquivo:


```diff
diff --git a/src/main.rs b/src/main.rs
index 85bc4bd..1d5c7a2 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -2,7 +2,6 @@ use crate::parser::parse_html_to_lei;
 use std::collections::HashMap;
 use std::env;
 use std::fs::File;
-use std::io::Write;
 use walkdir::{DirEntry, WalkDir};
 
 mod parser;
@@ -43,11 +42,8 @@ fn main() {
         println!("diretório {}: {} arquivos lidos", directory, files_number);
     }
 
-    let data = leis.join(",");
-    let mut leis_file = File::create("leis.json").expect("Unable to create file");
-    leis_file
-        .write_all(format!("[{}]", data).as_bytes())
-        .expect("Unable to write data");
+    let leis_file = File::create("leis.json").expect("Unable to create file");
+    serde_json::to_writer(leis_file, &leis).expect("Unable to write data");
 }
 
 fn is_not_hidden(entry: &DirEntry) -> bool {

```

Reply:
Pow, muito bom. Essa do array eu procurei procurei e não consegui achar como fazia. Eu colocava o array no `println!` e saia bonitinho, mas não na hora de escrever no arquivo. Provavelmente na hora esqueci da `Display` trait.

Reply:
Fazendo um segundo comentário aqui porque ficou show demais com o `to_writer` <3